components/ai_rewriter/renderer/ai_rewriter_button_positioner.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_rewriter/renderer/ai_rewriter_button_positioner.h"
+
+#include "base/check.h"
+#include "base/functional/bind.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/renderer/render_frame.h"
+#include "content/public/renderer/render_frame_observer.h"
+#include "third_party/blink/public/common/associated_interfaces/associated_interface_provider.h"
+#include "third_party/blink/public/web/web_document.h"
+#include "third_party/blink/public/web/web_dom_event.h"
+#include "third_party/blink/public/web/web_element.h"
+#include "third_party/blink/public/web/web_frame_widget.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+#include "third_party/blink/public/web/web_local_frame_client.h"
+#include "third_party/blink/public/web/web_node.h"
+#include "third_party/blink/public/web/web_range.h"
+#include "third_party/blink/public/web/web_view.h"
+#include "ui/gfx/geometry/rect.h"
+
+namespace ai_rewriter {
+
+AIRewriterButtonPositioner::AIRewriterButtonPositioner(
+    content::RenderFrame* frame)
+    : content::RenderFrameObserver(frame) {
+  frame->GetRemoteAssociatedInterfaces()->GetInterface(&button_);
+}
+
+AIRewriterButtonPositioner::~AIRewriterButtonPositioner() = default;
+
+void AIRewriterButtonPositioner::OnDestruct() {
+  delete this;
+}
+
+void AIRewriterButtonPositioner::DidCreateDocumentElement() {
+  auto document = render_frame()->GetWebFrame()->GetDocument();
+  CHECK(!document.IsNull());
+  remove_listener_ = document.AddEventListener(
+      blink::WebNode::EventType::kSelectionchange,
```

Comment:
can this event be triggered by JS in a background tab? will everything work the way you expect it? We don't want the button to appear randomly if unrelated tab decide to do some random stuff.

Reply:
It can be triggered by the page:
[Screencast from 2024-08-27 09-04-22.webm](https://github.com/user-attachments/assets/3d8c4609-3bbf-4701-abbd-798ca498a560)

But at the same time, I think thats fine because that text **is** selected, which is what we want to trigger the popup. It can't be shown outside the bounds of the window, so its similar to any background change to the page.

WDYT?

Reply:
> But at the same time, I think thats fine because that text is selected, which is what we want to trigger the popup. It can't be shown outside the bounds of the window, so its similar to any background change to the page.

You should at least check for a user gesture. Consider a html/js benchmark that tests the text selection metrics.

Reply:
this is what's wrong with the current impl:

https://github.com/user-attachments/assets/afe20730-bc7a-4292-862c-ed55556f401f

Reply:
this also does not look great.

1. why the button disappears and flashes?
2. why the popup window is created this far away?

https://github.com/user-attachments/assets/5141728e-cbcc-4c49-9dda-aa4ff2725df8

Reply:
The second one seems to be because internally there are a bunch of elements and the `WebRange` from `GetSelection` is empty - I've removed the length check in the button positioner.

I've added a check that the current WebContents is active - this is only for tabs in the same window, so I think it should fix it? If you'd rather I go the UserGesture route, how do I check for that at the renderer level?

Reply:
The popup being in the wrong place is interesting - what platform are you on?

Reply:
> The popup being in the wrong place is interesting - what platform are you on?

this was on windows

Reply:
> I've added a check that the current WebContents is active - this is only for tabs in the same window, so I think it should fix it? If you'd rather I go the UserGesture route, how do I check for that at the renderer level?

both checks should exist. UserGesture is tested via `blink::Frame::HasTransientUserActivation`. see `third_party/blink/public/common/frame/user_activation_state.h` and https://mustaqahmed.github.io/user-activation-v2/ for more details.

Reply:
Sweet, thanks @goodov - that was easier than I was expecting! I've updated the positioning code - I think the issue was I wasn't using `views::View::ConvertPointToWidget`